### DISCUSSION TITLE
A hint of member starts with 2 underlines was given out of its class

### DISCUSSION THREAD

**Discussion by zhang-stephen** [microsoft/pylance-release#4258] [Enhancement]
## Environment data

-   Language Server version: 2023.4.11
-   OS and version: Windows 11
-   Python version (& distribution if applicable, e.g. Anaconda): Python 3.8.10

## Code Snippet

```python
class Node:
    def __init__(self):
        self.__traversed = False

    def is_traversed(self) -> bool: return self.__traversed

    def traversed(self): self.__traversed = True

n = Node()
n.__traversed = True
```

## Repro Steps

When the last line was being typed, a hint would appear as the screenshot:
![image](https://user-images.githubusercontent.com/17326718/233028404-1e66fc34-7e3e-41e5-b5f0-6dcc02934050.png)

## Expected behavior

do not give hints about members those start with 2 underlines out of their classes, they are 'private' for invoker of these classes.

## Actual behavior

The hint as the screenshot.

## Logs

*no logs for this issue.*

**Comment by rchiodo** [comment]
Thanks for the issue but this is actually working as designed. You may or may not want to call private members. They're sorted last as they wouldn't normally be used.

There's a 'type checking mode' that will catch these usages though:

```
"python.analysis.typeCheckingMode": "strict"
```

![image](https://user-images.githubusercontent.com/19672699/233141637-c6b86e19-a40b-4d6c-ad39-cd14a05762ca.png)

Not showing them in completions would be an enhancement request.

**Comment by erictraut** [comment]
Private class member names are "mangled" by the Python compiler, so they are not accessible from outside the class, at least by their unmangled name.

```python
class Foo:
    __bar: int = 1

f = Foo()

print(f.__bar)  # AttributeError: 'Foo' object has no attribute '__bar'
```

So if the completion provider were to implement this change, it would need to be smart enough to know whether the access is performed inside or outside of the owning class. As you can see from the screen shot above, pyright does distinguish between these two cases when performing its strict-mode `reportPrivateUsage` check.

**Comment by judej** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.

**Reply by jbhfield** [reply]
Sorry to grave dig. Where would I show my support for a feature like this? I frequently run into this type of thing with Python, and it would clean up the VSCode suggestions immensely, even though there is no true concept of privacy in Python like in other languages.

**Reply by Andrej730** [reply]
@jbhfield usually by upvoting the first post in this discussion.
